daemon/list.go

```diff
@@ -90,6 +90,41 @@ func (daemon *Daemon) Containers(config *types.ContainerListOptions) ([]*types.C
 	return daemon.reduceContainers(config, daemon.transformContainer)
 }
 
+func (daemon *Daemon) filterByNameIDMatches(ctx *listContext) []*container.Container {
+	names := ctx.filters.Get("name")
+	ids := ctx.filters.Get("id")
+	if len(names)+len(ids) == 0 {
+		// if name or ID filters are not in use, return to
+		// standard behavior of walking the entire container
+		// list from the daemon's in-memory store
+		return daemon.List()
+	}
+
+	matches := make(map[string]bool)
+	// look for name matches; errors only report not found and can be ignored
+	if len(names) > 0 {
+		for id, idNames := range ctx.names {
+			for _, eachName := range idNames {
```

Comment:
Hmm, yes, for filters that include both ID and name, it would make sense to limit the name matching to the subset of containers which met the ID filter terms. Will make this loop a bit more complex, but for a large # of containers on a daemon which use both search terms, it would potentially be significant. On the other hand combining filter of ID and name both seems unusual, but supported, so I guess it makes sense to do it efficiently! :)